tests/integration/test_with_portainer_service.py

```diff
@@ -21,9 +65,230 @@ async def test_portainer_connection(
+        aiohttp_client_session,
+        portainer_baerer_code,
+        portainer_endpoint_id,
+    )
+    current_stack_name = "pytestintegration"
+    # Assuring a clean state by deleting any remnants
+    os.system("docker stack rm " + current_stack_name)
+
+    new_stack = await portainer.post_new_stack(
+        base_url=portainer_url,
+        app_session=aiohttp_client_session,
+        bearer_code=portainer_baerer_code,
+        swarm_id=swarm_id,
+        endpoint_id=portainer_endpoint_id,
+        stack_name=current_stack_name,
+        stack_cfg=valid_docker_stack,
+    )
+    time.sleep(2)
+    stack_id = await portainer.get_current_stack_id(
+        base_url=portainer_url,
+        app_session=aiohttp_client_session,
+        bearer_code=portainer_baerer_code,
+        stack_name=current_stack_name,
+    )
+    time.sleep(2)
+    await portainer.delete_stack(
+        base_url=portainer_url,
+        app_session=aiohttp_client_session,
+        bearer_code=portainer_baerer_code,
+        stack_id=int(stack_id),
+        endpoint_id=portainer_endpoint_id,
+    )
+
+
+async def test_portainer_redeploys_when_sha_of_tag_in_docker_registry_changed(
+    loop: asyncio.AbstractEventLoop,
+    portainer_container: tuple[URL, str],
+    aiohttp_client_session: ClientSession,
+    portainer_baerer_code: str,
+    portainer_endpoint_id: int,
+    valid_docker_stack_with_local_registry: dict[str, Any],
+    docker_registry: str,
+    docker_registry_image_injector: Callable,
+    faker: Faker,
+    docker_swarm: None,
+):
+    ### Push image to local registry
+    # Note for the future: This boilerplate might also help: https://github.com/docker/docker-py/issues/2104#issuecomment-410802929
+    client = docker.from_env()
+    img_name = "itisfoundation/sleeper"
+    img_tag = "2.1.1"
+
+    # Note DK2023: On my wsl2 machine, the image is pulled as 127.0.0.1:5000/simcore/services/comp/itis/sleeper:2.1.1,
+    sleeper_service = docker_registry_image_injector(img_name, img_tag, faker.email())
+    # Retag image
+    sleeper_image_name = f"{docker_registry}/{sleeper_service['image']['name']}:{sleeper_service['image']['tag']}"
+    sleeper_image = client.images.get(sleeper_image_name)
+    ###
+    portainer_url, _ = portainer_container
+    swarm_id = await portainer.get_swarm_id(
+        portainer_url,
+        aiohttp_client_session,
+        portainer_baerer_code,
+        portainer_endpoint_id,
+    )
+    # Assuring a clean state by deleting any remnants
+    current_stack_name = "pytestintegration"
+
+    os.system("docker stack rm " + current_stack_name)
+
+    new_stack = await portainer.post_new_stack(
+        base_url=portainer_url,
+        app_session=aiohttp_client_session,
+        bearer_code=portainer_baerer_code,
+        swarm_id=swarm_id,
+        endpoint_id=portainer_endpoint_id,
+        stack_name=current_stack_name,
+        stack_cfg=valid_docker_stack_with_local_registry,
+    )
+    time.sleep(2)
```

Comment:
TIP: try to avoid fixed delays when possible.

Your approach has
PROS: easy to implement
CONS: 
- typically make tests longer than needed
- constants delay is a heuristic that might change with time

An alternative is to observe frequently a given state: Create a check function that you retry (i.e.  with ``tenacity``) very frequently until it validates